pkg/base1/test-channel.ts

```diff
@@ -0,0 +1,538 @@
+import QUnit from "qunit-tests";
+
+import { transport_globals } from 'cockpit/_internal/transport';
+import { Channel, ChannelControlMessage, ChannelOptions, ChannelPayload } from 'cockpit/channel';
+
+// simple function to 'use' values (to avoid 'unused' warnings)
+function use(...args: unknown[]) {
+    console.log(args.length);
+}
+
+// Many of the functions in this file are not designed to run, but to be
+// typechecked.  Here's some assertions for the types we'll check.
+function is_str(x: string): void {
+    use(x);
+}
+
+function is_bytes(x: Uint8Array): void {
+    use(x);
+}
+
+// verify that type assertions are fundamentally working the way we think
+function verify_type_assertions() {
+    const bytes = new Uint8Array();
+    const str = "";
+
+    // These are positive assertions
+    is_str(str);
+    is_bytes(bytes);
+
+    // These are typing fails.  tsc will verify that they fail.
+    // @ts-expect-error Cannot pass bytes to is_str()
+    is_str(bytes);
+
+    // @ts-expect-error Cannot pass string to is_bytes()
+    is_bytes(str);
+}
+
+function is_str_channel(x: Channel<string>): void {
+    x.on('data', msg => is_str(msg));
+
+    // @ts-expect-error: of course string channels don't get binary data...
+    x.on('data', msg => is_bytes(msg));
+}
+
+function is_bytes_channel(x: Channel<Uint8Array>): void {
+    x.on('data', msg => is_bytes(msg));
+
+    // @ts-expect-error: of course binary channels don't get string data...
+    x.on('data', msg => is_str(msg));
+}
+
+// tsc seems to have inconsistent treatment of `void` argument types when
+// things get complicated.  If this ever gets fixed, search for 'VOID BUG'
+// and adjust each test
+
+// TODO: file that bug...
```

Comment:
TODO: write review comment

Reply:
thanks :)

Reply:
Already exists: https://github.com/microsoft/TypeScript/issues/29131

Reply:
Added a comment with a link to that issue.